dali/core/mm/cuda_vm_resource_test.cc

```diff
@@ -0,0 +1,305 @@
+    }
+  }
+
+  void TestRegionExtendAfter() {
+    cuda_vm_resource res;
+    res.block_size_ = 4 << 20;  // 4 MiB
+    const int b1 = 32, b2 = 64;
+    const size_t s1 = b1 * res.block_size_;
+    const size_t s2 = b2 * res.block_size_;
+    res.va_ranges_.push_back(cuvm::CUMemAddressRange::Reserve(s1 + s2));
+    cuvm::CUAddressRange total = res.va_ranges_.back();
+    cuvm::CUAddressRange part1 = { total.ptr(),           s1 };
+    cuvm::CUAddressRange part2 = { total.ptr() + s1,      s2 };
+    res.va_add_region(part1);
+    ASSERT_EQ(res.va_regions_.size(), 1u);
+    MapRandomBlocks(res.va_regions_[0], 10);
+    va_region_backup va1 = Backup(res.va_regions_[0]);
+    res.va_add_region(part2);
+    ASSERT_EQ(res.va_regions_.size(), 1u);
+    auto &region = res.va_regions_.back();
+    ASSERT_EQ(region.num_blocks(), b1 + b2);
+    ComparePart(region, 0, va1);
+    for (int i = b1; i < b1 + b2; i++) {
+      EXPECT_FALSE(region.mapped[i]);
+      EXPECT_FALSE(region.mapped[i]);
+    }
+  }
+
+  void TestRegionExtendBefore() {
+    cuda_vm_resource res;
+    res.block_size_ = 4 << 20;  // 4 MiB
+    const int b1 = 32, b2 = 64;
+    const size_t s1 = b1 * res.block_size_;
+    const size_t s2 = b2 * res.block_size_;
+    res.va_ranges_.push_back(cuvm::CUMemAddressRange::Reserve(s1 + s2));
+    cuvm::CUAddressRange total = res.va_ranges_.back();
+    cuvm::CUAddressRange part1 = { total.ptr(),           s1 };
+    cuvm::CUAddressRange part2 = { total.ptr() + s1,      s2 };
+    res.va_add_region(part2);
+    ASSERT_EQ(res.va_regions_.size(), 1u);
+    MapRandomBlocks(res.va_regions_[0], 10);
+    va_region_backup va1 = Backup(res.va_regions_[0]);
+    res.va_add_region(part1);
+    ASSERT_EQ(res.va_regions_.size(), 1u);
+    auto &region = res.va_regions_.back();
+    ASSERT_EQ(region.num_blocks(), b1 + b2);
+    ComparePart(region, b1, va1);
+    for (int i = 0; i < b1; i++) {
+      EXPECT_FALSE(region.mapped[i]);
+      EXPECT_FALSE(region.mapped[i]);
+    }
+  }
+
+  void TestRegionMerge() {
+    cuda_vm_resource res;
+    res.block_size_ = 4 << 20;  // 4 MiB
+    const int b1 = 32, b2 = 64, b3 = 32;
+    const size_t s1 = b1 * res.block_size_;
+    const size_t s2 = b2 * res.block_size_;
+    const size_t s3 = b3 * res.block_size_;
+    res.va_ranges_.push_back(cuvm::CUMemAddressRange::Reserve(s1 + s2 + s3));
+    cuvm::CUAddressRange total = res.va_ranges_.back();
+    cuvm::CUAddressRange part1 = { total.ptr(),           s1 };
+    cuvm::CUAddressRange part2 = { total.ptr() + s1,      s2 };
+    cuvm::CUAddressRange part3 = { total.ptr() + s1 + s2, s3 };
+    res.va_add_region(part1);
+    ASSERT_EQ(res.va_regions_.size(), 1u);
+    MapRandomBlocks(res.va_regions_[0], 10);
+    va_region_backup va1 = Backup(res.va_regions_[0]);
+    res.va_add_region(part3);
+    ASSERT_EQ(res.va_regions_.size(), 2u);
+    MapRandomBlocks(res.va_regions_[1], 12);
+    va_region_backup va3 = Backup(res.va_regions_[1]);
+    res.va_add_region(part2);
```

Comment:
you don't call MapRandomBlocks on part2?

Reply:
Yes and that's intentional - I want to check that this part is inserted as empty into the resulting region.